Ignore some objects in LeakSanitizer.

samth opened the pull request:
Really this should ignore fewer than this, but it's not yet clear
where the calls should go.

Related to #3839.

samth commented:
This is effectively doing the same as having a suppression file for these function calls, which isn't really what we want. But I don't know where the right place to make calls to `__lsan_ignore_object()` is, especially since I think the objects are "leaked" inside Racket code.

samth commented:
Another problem is that building with ASan doesn't work on an recent ubuntu install; I had to do this development inside a docker container using the `racket/racket-ci` docker image.

mflatt commented:
The fundamental misatch is related to roots, so I'd expect the right change to be using `__lsan_register_root_region` (maybe just before exiting) in Chez Scheme, not changing rktio.

samth commented:
Where would I call that (and with what sorts of arguments)?

mflatt commented:
I had imagined adding an exit callback that registers all GC pages. But on further reflection, pointers to C objects are likely to be represented as fixnums or bignums in a way that won't look like a pointer to LeakSanitizer.

So, maybe a better strategy is to have some way (maybe an environment variable) to change exit to force a custodian shutdown. Then, we'd have to deal with whatever is left.

samth committed:
rktio: add working rktio_lsan_ignore_object + LSan suppressions

Reworks PR #3853. The original rktio_lsan_ignore_object did not compile
(malformed printf) and was a no-op. This adds a correct implementation
(guarded for GCC __SANITIZE_ADDRESS__ and Clang __has_feature) and calls
it on the rktio converter, which together with the iconv state reachable
through it is retained for the process lifetime via the Racket/Chez GC
heap (which LeakSanitizer does not scan) and so is otherwise reported as
a ~33KB false-positive leak.

Also adds lsan-suppressions.supp covering the remaining intentional
process-lifetime allocations (Chez mutex/condition/boot, rktio std fds,
long-term poll set, current-directory). With an --enable-asan build,
ASAN_OPTIONS=detect_leaks=1 LSAN_OPTIONS=suppressions=lsan-suppressions.supp
gives a clean leak report while still catching real leaks.
